src/System.Linq/tests/FirstTests.cs

```diff
@@ -0,0 +1,187 @@
+                    where x > Int32.MinValue
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        [Fact]
+        public void SameResultsRepeatCallsStringQuery()
+        {
+            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
+                    where !String.IsNullOrEmpty(x)
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        public void TestEmptyIList<T>()
+        {
+            T[] source = { };
+            
+            Assert.NotNull(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyIListT()
+        {
+            TestEmptyIList<int>();
+            TestEmptyIList<string>();
+            TestEmptyIList<DateTime>();
+            TestEmptyIList<FirstTests>();
+        }
+
+        [Fact]
+        public void IListTOneElement()
+        {
+            int[] source = { 5 };
+            int expected = 5;
+
+            Assert.NotNull(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsDefault()
+        {
+            int?[] source = { null, -10, 2, 4, 3, 0, 2 };
+            int? expected = null;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsNotDefault()
+        {
+            int?[] source = { 19, null, -10, 2, 4, 3, 0, 2 };
+            int? expected = 19;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        private static IEnumerable<T> EmptySource<T>()
+        {
+            yield break;
+        }
+
+        private static void TestEmptyNotIList<T>()
+        {
+            var source = EmptySource<T>();
+
+            Assert.Null(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyNotIListT()
+        {
+            TestEmptyNotIList<int>();
+            TestEmptyNotIList<string>();
+            TestEmptyNotIList<DateTime>();
+            TestEmptyNotIList<FirstTests>();
+        }
+
+        [Fact]
+        public void OneElementNotIListT()
+        {
+            IEnumerable<int> source = NumList(-5, 1);
+            int expected = -5;
+
+            Assert.Null(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void ManyElementsNotIListT()
+        {
+            IEnumerable<int> source = NumList(3, 10);
+            int expected = 3;
+
+            Assert.Null(source as IList<int>);
```

Comment:
True, but an improvement that gets rid of NumList entirely would slip past that.

Indeed, I considered replacing NumList with `Enumerable.Range(-5, 1)` as being the most natural way of doing the same thing, and then I considered that I myself have suggested that that should return something that implements `IList<int>`, so I was considering a change that combined with something I myself suggested would break the purpose of the test, and decided that maybe that check isn't so silly after all.